server/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java

```diff
@@ -16,182 +16,109 @@
-            in -> new ConcreteShardRequest<>(requestReader, in), this::handlePrimaryRequest);
+        this.transportOptions = transportOptions;
+        this.syncGlobalCheckpointAfterOperation = syncGlobalCheckpointAfterOperation;
 
         // we must never reject on because of thread pool capacity on replicas
-        transportService.registerRequestHandler(transportReplicaAction, executor, true, true,
-            in -> new ConcreteReplicaRequest<>(replicaRequestReader, in), this::handleReplicaRequest);
-
-        this.transportOptions = transportOptions(settings);
-
-        this.syncGlobalCheckpointAfterOperation = syncGlobalCheckpointAfterOperation;
+        transportService.registerRequestHandler(actionName + "[r]", executor, true, true,
+            in -> new ConcreteReplicaRequest<>(replicaRequestReader, in),
+            (replicaRequest, channel, task) -> execute(task, replicaRequest,
+                new ChannelActionListener<>(channel, this.actionName, replicaRequest)));
     }
 
     @Override
-    protected void doExecute(Task task, Request request, ActionListener<Response> listener) {
-        assert request.shardId() != null : "request shardId must be set";
-        new ReroutePhase((ReplicationTask) task, request, listener).run();
-    }
-
-    protected ReplicationOperation.Replicas<ReplicaRequest> newReplicasProxy() {
-        return new ReplicasProxy();
+    protected void doExecute(Task task, ConcreteReplicaRequest<ReplicaRequest> request,
+                             ActionListener<ReplicaResponse> listener) {
+        new AsyncReplicaAction(request, listener, (ReplicationTask) task).run();
     }
 
-    protected abstract Response newResponseInstance();
-
-    /**
-     * Resolves derived values in the request. For example, the target shard id of the incoming request, if not set at request construction.
-     * Additional processing or validation of the request should be done here.
-     *
-     * @param indexMetaData index metadata of the concrete index this request is going to operate on
-     * @param request       the request to resolve
-     */
-    protected void resolveRequest(final IndexMetaData indexMetaData, final Request request) {
-        if (request.waitForActiveShards() == ActiveShardCount.DEFAULT) {
-            // if the wait for active shard count has not been set in the request,
-            // resolve it from the index settings
-            request.waitForActiveShards(indexMetaData.getWaitForActiveShards());
-        }
-    }
-
-    /**
-     * Primary operation on node with primary copy.
-     *
-     * @param shardRequest the request to the primary shard
-     * @param primary      the primary shard to perform the operation on
-     */
-    protected abstract void shardOperationOnPrimary(Request shardRequest, IndexShard primary,
-        ActionListener<PrimaryResult<ReplicaRequest, Response>> listener);
+    protected abstract void acquireReplicaOperationPermit(IndexShard replica,
```

Comment:
Perhaps we should just add a boolean to this method to denote whether one or all permits should be acquired, we can then avoid exposing this method to subclasses and the extra delegation required here.